jobs/ltv_daily.py

```diff
@@ -0,0 +1,206 @@
+    return combined_counts, bgf
+
+
+def catch_none(x):
+    if x == None:
+        return 0
+    return x
+
+
+def ltv_predict(t, frequency, recency, T, model):
+    pred = model.conditional_expected_number_of_purchases_up_to_time(
+        t, catch_none(frequency), catch_none(recency), catch_none(T)
+    )
+
+    if pred > t:
+        return float(t)
+    return float(pred)
+
+
+def main(
+    submission_date,
+    project_id,
+    dataset_id,
+    source_qualified_table_id,
+    intermediate_table_id,
+    model_input_table_id,
+    model_output_table_id,
+    temporary_gcs_bucket,
+):
+    """Model the lifetime-value (LTV) of clients based on search activity.
+
+    This reads a single partition from a source table into an intermediate table
+    in an analysis dataset. The table is transformed for modeling. The model
+    inputs and predictions are then stored into separate tables in the same
+    dataset.
+    """
+    print(f"Running ltv_daily job for {submission_date}")
+
+    bq = bigquery.Client()
+    table_ref = bq.dataset(dataset_id, project=project_id).table(intermediate_table_id)
+
+    # define the job configuration for the query
+    # set params and output destination for the materialized
+    # dataset
+    job_config = bigquery.QueryJobConfig()
+    job_config.query_parameters = [
+        bigquery.ScalarQueryParameter("submission_date", "STRING", submission_date)
+    ]
+    job_config.destination = table_ref
+    job_config.write_disposition = bigquery.WriteDisposition.WRITE_TRUNCATE
+
+    query = f"""
+	SELECT
+	    *
+	FROM
+	    `{source_qualified_table_id}`
+	WHERE
+	    submission_date = @submission_date
+	"""
+    query_job = bq.query(query, job_config=job_config)
+    query_job.result()
+
+    spark = SparkSession.builder.getOrCreate()
+    search_rfm_full = (
+        spark.read.format("bigquery")
+        .option("table", f"{project_id}.{dataset_id}.{intermediate_table_id}")
+        .load()
+    )
+
+    columns = [
+        [
+            F.col(str(metric + ".frequency")).alias(metric + "_frequency"),
+            F.col(str(metric + ".recency")).alias(metric + "_recency"),
+            F.col(str(metric + ".T")).alias(metric + "_T"),
+        ]
+        for metric in ["days_searched", "days_tagged_searched", "days_clicked_ads"]
+    ]
+
+    # flatten list
+    columns = [item for sublist in columns for item in sublist]
+
+    model_perf_data = pd.DataFrame()
+    model_pred_data = None
+    pred_metrics = ["days_searched", "days_tagged_searched", "days_clicked_ads"]
+    search_rfm_ds = search_rfm_full.limit(TRAINING_SAMPLE).select(columns).toPandas()
+    for metric in pred_metrics:
+        print(metric)
+        # train and extract model performace
+        model_perf, model = train_metric(search_rfm_ds, metric, plot=False, penalty=0.8)
+        model_perf["pct"] = model_perf.Model / (model_perf.Actual + 1) - 1
+        model_perf["metric"] = metric
+        model_perf["date"] = submission_date
```

Comment:
@benmiroglio This job had `d` as the variable that was set here. Is `submission_date` the correct variable to use?

Reply:
yes, this should be `submission_date`